### PR TITLE
TextureDecoder: Fix off-by-one errors in CMPR

### PR DESCRIPTION
**Pull request by hthh**
This fixes some (probably imperceptible) issues in CMPR decoding, based on hardware testing.

This code previously showed different output on Wii and Dolphin, but with this PR it appears consistent: https://gist.github.com/hthh/7968c721f8296e8ea57477d15a2d2aee (note: requires real XFB). I had some problems with networking, so I haven't integrated my test code with the hwtests yet. I'll try to do that soon.

### CODE REVIEW

File: Source/Core/VideoCommon/TextureDecoder_Common.cpp [hunk 1]
```diff
@@ -434,6 +434,11 @@ static inline u32 MakeRGBA(int r, int g, int b, int a)
   return (a << 24) | (b << 16) | (g << 8) | r;
 }
 
+static inline u8 DXTBlend(u8 v1, u8 v2)
```
Comment (lioncash) [inline Source/Core/VideoCommon/TextureDecoder_Common.cpp]: This can be `static constexpr` or `static` instead of `static inline`. `inline` is generally unnecessary in the presence of static. Ditto for other one in the generic decoder.

File: Source/Core/VideoCommon/TextureDecoder_Common.cpp [hunk 2]
```diff
@@ -434,6 +434,11 @@ static inline u32 MakeRGBA(int r, int g, int b, int a)
...
+{
+  return ((v1 * 3 + v2 * 5) >> 3);
```
Comment (lioncash) [inline Source/Core/VideoCommon/TextureDecoder_Common.cpp]: this needs an explicit cast to `u8` otherwise it truncates implicitly (arithmetic would promote to int).

```cpp
...
return static_cast<u8>((v1 * 3 + v2 * 5) >> 3);
...
```

Comment (hthh) [inline Source/Core/VideoCommon/TextureDecoder_Common.cpp]: Since the arguments are also implicitly truncated each time this is called, I'll probably just change all the types to "int" and avoid both conversions (when GitHub starts feeling better). Does that work?

Comment (lioncash) [inline Source/Core/VideoCommon/TextureDecoder_Common.cpp]: yeah, that's fine

File: Source/Core/VideoCommon/TextureDecoder_Common.cpp [hunk 3]
```diff
@@ -434,6 +434,11 @@ static inline u32 MakeRGBA(int r, int g, int b, int a)
   return (a << 24) | (b << 16) | (g << 8) | r;
 }
 
+static constexpr int DXTBlend(int v1, int v2)
```
Comment (iwubcode) [inline Source/Core/VideoCommon/TextureDecoder_Common.cpp]: Maybe unnecessary but should we put this in a TextureDecoder_Util that both this and "Generic" use?  MakeRGBA/DXTBlock should probably also go there.  This should probably be a separate PR...

Comment (iwubcode) [inline Source/Core/VideoCommon/TextureDecoder_Common.cpp]: Can you either comment here (or add a comment?) on how you calculated this value?

**Comment by JosJuice** [comment]
They're not pure black, they're just very close to black, meaning the difference in colors is very small. If you enable Bright mode, you'll see where the differences are.

**Comment by iwubcode** [comment]
@JosJuice  - ah ok, I wondered what that option was for.  Thanks!

**Comment by Parlane** [comment]
Cant you just put DXTBlend in TextureDecoder.h ?

**Comment by iwubcode** [comment]
@Parlane - yes, this makes sense.  Sorry, I didn't look at the code base before I made my suggestion.

Random question - is there a reason that we've named the cpp for TextureDecoder something different (at least I'm assuming that is what TextureDecoder_Common.cpp is?)

File: Source/Core/VideoCommon/TextureDecoder_Common.cpp (hunk 3 above)
Comment (hthh) [inline Source/Core/VideoCommon/TextureDecoder_Common.cpp]: The expression comes from guessing and checking, using a script to compare the results of my guess to values obtained from the hardware.

I'll get rid of the `DXTBlend` duplication.

Comment (iwubcode) [inline Source/Core/VideoCommon/TextureDecoder_Common.cpp]: Very cool.  One other question - can I ask how you get the results from hardware?  Sorry for all the questions, I just like insight on how developers work!

**Comment by lioncash** [comment]
@Parlane It should probably just go in a new utility header, so it doesn't conflate the interface with implementation details.

File: Source/Core/VideoCommon/TextureDecoder_Common.cpp (hunk 3 above)
Comment (iwubcode) [inline Source/Core/VideoCommon/TextureDecoder_Common.cpp]: Just saw the c compare script, not sure how I missed it the first time!  Thanks @hthh !

File: Source/Core/VideoCommon/TextureDecoder.h [hunk 4]
```diff
@@ -88,3 +88,8 @@ void TexDecoder_SetTexFmtOverlayOptions(bool enable, bool center);
 /* Internal method, implemented by TextureDecoder_Generic and TextureDecoder_x64. */
 void _TexDecoder_DecodeImpl(u32* dst, const u8* src, int width, int height, int texformat,
                             const u8* tlut, TlutFormat tlutfmt);
+
+static constexpr int DXTBlend(int v1, int v2)
```
Comment (lioncash) [inline Source/Core/VideoCommon/TextureDecoder.h]: Since this is in a header, `static` should be dropped (constexpr implies inline, so there's no need to specify inline to resolve linkage in this case).

**Comment by iwubcode** [comment]
@lioncash - that was my original thought too but can't the implementation details be in common?  Isn't that where the other defined functions live?

**Comment by lioncash** [comment]
sure, either way works.

File: Source/Core/VideoCommon/TextureDecoder_Common.cpp (hunk 3 above)
Comment (hthh) [inline Source/Core/VideoCommon/TextureDecoder_Common.cpp]: I wrote a Wii homebrew that displayed the hex values from decoding on the screen (see the gist link in the PR description). Then I ran it, took a photo of my TV, and then typed a number of test values in by hand and checked them twice.

You can have a look at how Dolphin's real hardware tests work: https://github.com/dolphin-emu/hwtests - I had problems with the network stuff, so I used the libogc example "template.c", which prints to the screen. I'll hopefully work out the network issues to make it easier next time.

File: Source/Core/VideoCommon/TextureDecoder_Generic.cpp [hunk 5]
```diff
@@ -170,20 +159,17 @@ static void DecodeDXTBlock(u32* dst, const DXTBlock* src, int pitch)
   if (c1 > c2)
   {
     // Approximation of x/3: 3/8 (1/2 - 1/8)
```
Comment (stenzek) [inline Source/Core/VideoCommon/TextureDecoder_Generic.cpp]: This comment seems to be invalid now (as you now have 3/8 and 5/8 in the Util file).

**Comment by JMC47** [comment]
This does have some very, very minor, but perceptible differences in Burnout 2's truck textures.  Cool.
